### ISSUE TITLE
Error when declaring member properties outside of the constructor

### ISSUE THREAD

**Issue by ryandegruyter** [ngrx/effects#17]
export class MyEffects{
 private _update$: StateUpdates<any>; 
 constructor(updates$: StateUpdates<any>) { this._updates$ = updates$ }

  @Effect()
  ....
}

When I use the above code I get the following exception:
`EXCEPTION: Error during instantiation of Token @ngrx/effects Bootstrap Effects! (Token Application Initializer -> Token @ngrx/effects Bootstrap Effects).`

But when I declare my class properties inside the constructor parameter list, it works fine.
export class MyEffects{

 constructor(private updates$: StateUpdates<any>) {}

  @Effect()
  ....
}

any clues?

**Comment by dsebastien** [comment]
We're also facing this issue, have you found anything so far?

**Comment by ryandegruyter** [comment]
Unfortunatly no, I haven't looked into it any further. I ended up declaring them inside the constructor parameter list.

Are you using ionic or any angular 2 starter projects? I believe I was using ionic 2 beta 7 at the time.

**Comment by dsebastien** [comment]
Bummer.

No, we've adapted ngrx so that we could use it in our AngularJS applications (preparing for a later switch to ng2). I'll try this again with the beta of TypeScript 2 and I'll try to look at what differs in the emitted ES5 code...

**Comment by dsebastien** [comment]
We've dived into this today at work and found out the reason why this error occurs.
Basically the issue boils down to the way the TypeScript compiler adds the member fields on the generated JS code.

If you define stateUpdates without the constructor notation (private ...), then the field on which you place the @Effect decorator ends up before the stateUpdates one, leading to the stateUpdates reference being invalid (since of course when that line is executed, stateUpdates is not set yet.

One solution with the code as it stands is to do something along those lines:

```
class AuthEffects {
   private update$:StateUpdates<any>;
   
   @Effect
   private loginEffect:Observable<any>;

   constructor(stateUpdates:StateUpdates<any>, ...) {
      this.stateUpdates = stateUpdates;
      this.loginEffect = this.loginEffect();
   }

   public loginEffect():Observable<any> {
      return this.updates$
      .whenAction('LOGIN')
      ...
}
```

With the above, the variables will be defined in the correct order in the emitted JS code.

Although we didn't find this all too elegant (more lines :x), which is why we've adapted the code of ngrx effects so that the @Effect decorator could be set on a function rather than on a field. We'll send a PR with this change soon.

With our proposed change, you can just write the above like this:

```
class AuthEffects {
   private update$:StateUpdates<any>;

   constructor(stateUpdates:StateUpdates<any>, ...) {
      this.stateUpdates = stateUpdates;
   }
   
   @Effect
   public loginEffect():Observable<any> {
      return this.updates$
      .whenAction('LOGIN')
      ...
}
```

Which is IMO more readable, clearer and will cause less headaches.

**Comment by christophercr** [comment]
Hi, indeed, as @dsebastien mentioned, we managed to solve this issue by adding the support for functions to be decorated with @Effect. Obviously such functions should return Observable.

As a result, it is now possible to declare member properties outside the constructor, leading up to more readable and cleaner code ;)

This is the PR: https://github.com/ngrx/effects/pull/28

**Comment by MikeRyan52** [comment]
Fixed in v1.1.0
